### PR TITLE
crimson/os/seastore/cache: improve committing allocations by backref entries

### PR DESCRIPTION
**Pull request by cyx1231st**
This PR is supposed to be cleanup only, reworking https://github.com/ceph/ceph/pull/59775.

@zhscn @xxhdx1985126 

For the specific issue https://tracker.ceph.com/issues/66941, seems to me propagating remapped extents upon `prepare_record()` (PR59775) cannot address the issue, because the backref entries of fresh extents are still updated upon `complete_commit()`, and they have the same issue. And they cannot be simply moved to `prepare_record()` because their paddrs are unknown at that time.

So, IIUC a new segment state "COMMITTED" is still necessary to fix reclaiming, for both inline and ool segments.

~~TODO/Draft: still some cleanups to add.~~ Done.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
Changeset: fix compile issues in UT.

**Comment by cyx1231st** [comment]
jenkins test api

**Comment by cyx1231st** [comment]
jenkins test make check arm64

**Comment by cyx1231st** [comment]
Tests are all passing:
https://pulpito.ceph.com/yingxin-2024-12-23_02:45:56-crimson-rados-ci-yingxin-seastore-backref-cleanups-distro-default-smithi/

**Review by xxhdx1985126 — COMMENTED** [review]
LGTM otherwise:-)

File: src/crimson/os/seastore/cache.cc [hunk 1]
```diff
@@ -1708,7 +1708,7 @@ void Cache::complete_commit(
 	alloc_laddr = i->cast<lba_manager::btree::LBANode>()->get_node_meta().begin;
       } else {
 	assert(i->get_type() == extent_types_t::TEST_BLOCK_PHYSICAL);
-	alloc_laddr = L_ADDR_NULL;
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.cc]: This is allocation, so `L_ADDR_NULL` here isn't correct.

The commit (https://github.com/ceph/ceph/pull/61144/commits/665418e471bc60703b417591a80eeaa95fd287b2) fixes it to `L_ADDR_MIN`, is it reasonable to you?

**Review by xxhdx1985126 — APPROVED** [review]

**Comment by zhscn** [comment]
https://github.com/ceph/ceph/blob/58f7b6e07777228f30d6610f67fff55e1d3fbc41/src/crimson/os/seastore/cache.cc#L1824-L1828

The backref changes of existing blocks should also be moved to `Cache::prepare_recored()`, otherwise the GC trans would believe these remapped extents have been released completely.

**Comment by cyx1231st** [comment]
Discussed with @zhscn: remapping has 2 steps, retire and alloc, they must be committed atomically together, either in complete_commit() or prepare_record().

**Comment by cyx1231st** [comment]
Changeset: crimson/os/seastore/cache: move committing existing blocks to prepare_record()

@zhscn compared with the original PR, for existing extents, I also moved `add_extent()` related logic to `prepare_record()`.

**Comment by zhscn** [comment]
LGTM

**Comment by cyx1231st** [comment]
Changeset: fix a wrong assertion related to extent state.

**Comment by cyx1231st** [comment]
https://pulpito.ceph.com/yingxin-2024-12-26_01:55:39-crimson-rados-ci-yingxin-seastore-backref-cleanups-3-distro-default-smithi/
* 8053542: https://tracker.ceph.com/issues/68809

**Comment by cyx1231st** [comment]
jenkins test docs

**Pull request merged**